signac/contrib/job.py

```diff
@@ -24,31 +24,189 @@
-    to the contents of the file being modified this hook
-    calls :meth:`~Job._reset_sp` to rehash the state
-    point, compute a new job id, and move the folder.
-
-    Parameters
-    ----------
-    jobs : iterable of `Jobs`
-        List of jobs(instance of `Job`).
+class _StatePointDict(JSONDict):
+    """A JSON-backed dictionary for storing job state points.
 
+    There are three principal reasons for extending the base JSONDict:
+        1. Saving needs to trigger a job directory migration, and
+        2. State points are assumed to not support external modification, so
+           they never need to load from disk _except_ the very first time a job
+           is opened by id and they're not present in the cache.
+        3. It must be possible to load and/or save on demand during tasks like
+           job directory migrations.
     """
 
-    def __init__(self, *jobs):
-        self.jobs = list(jobs)
+    _PROTECTED_KEYS = ("_jobs",)
+
+    def __init__(
+        self,
+        jobs=None,
+        filename=None,
+        write_concern=False,
+        data=None,
+        parent=None,
+        *args,
+        **kwargs,
+    ):
+        # Multiple Python Job objects can share a single `_StatepointDict`
+        # instance because they are shallow copies referring to the same data
+        # on disk. We need to store these jobs in a shared list here so that
+        # shallow copies can point to the same place and trigger each other to
+        # update. This does not apply to independently created Job objects,
+        # even if they refer to the same disk data; this only applies to
+        # explicit shallow copies and unpickled objects within a session.
+        self._jobs = list(jobs)
+        super().__init__(
+            filename=filename,
+            write_concern=write_concern,
+            data=data,
+            parent=parent,
+            *args,
+            **kwargs,
+        )
 
-    def load(self):
+    def _load(self):
+        # State points never load from disk automatically. They are either
+        # initialized with provided data (e.g. from the state point cache), or
+        # they load from disk the first time state point data is requested for
+        # a Job opened by id (in which case the state point must first be
+        # validated manually).
         pass
 
-    def save(self):
-        """Reset the state point for all the jobs."""
-        for job in self.jobs:
-            job._reset_sp()
+    def _save(self):
+        # State point modification triggers job migration for all jobs sharing
+        # this state point (shallow copies of a single job).
+        new_id = calc_id(self)
+
+        # All elements of the job list are shallow copies of each other, so any
+        # one of them is representative.
+        job = next(iter(self._jobs))
+        old_id = job._id
+        if old_id == new_id:
+            return
+
+        tmp_statepoint_file = self.filename + "~"
```

Comment:
🤷‍♂️ 🙊 I really don't know what to do with this. We use `statepoint` in the code a lot. Probably leave this as-is...
```suggestion
        tmp_state_point_file = self.filename + "~"
```

Reply:
Yeah I intentionally chose to ignore variables with "statepoint" in them. We can do that in a separate PR on master at some point if we really want to, but that would also be API breaking (for example `project.open_job(state_point=...)`) so it would require additional discussion. As part of the signac 2.0 prototype Simon and I discussed moving to more generic terminology anyway (something like `attrs` rather than `statepoint`, which is very thermo-based), so this seems like a future discussion.